### DISCUSSION TITLE
Proposal: Type pattern for generic types

### DISCUSSION THREAD

**Discussion by gulshan** [dotnet/csharplang#268] [General]
Now, type patterns do not work with generic types. A type argument is required to determine actual type. Given the following generic classes (using `closed` and `record` syntax)-

```csharp
public closed abstract class Option<T> { }
public sealed class Some<T>(T Value) : Option<T> { }
public sealed class None<T>() : Option<T> { }
```

The code I want is write is-
```csharp
var x = SomeGenericMethod();
if(x is Some<int> i) // possible now, the type is needed to be mentioned explicitly
    Write(i.Value);
if(x is Some i) // proposing here. full type will be inferred
    Write(i.Value);
if(x is Some<var> i) // or this one if previous one is impossible
    Write(i.Value);
if(x is Some<_> i) // or this one
    Write(i.Value);
```

Grammar will be-

```antlr
pattern
    : type_pattern
    ;

type_pattern
    : type identifier
    | generic_type identifier
    ;
```

**Comment by YotaXP** [comment]
`if(x is Some i)` would not be possible, as it would be ambiguous between `Some<T>`, `Some<T,U>`, and a non-generic `Some` class.  The other options seem fine to me so far, though.

**Comment by orthoxerox** [comment]
@YotaXP why not allow it if the compiler can infer that there's only one applicable deconstructor?

upd: shouldn't've answered from my phone. I see it's not a deconstructor, it's a typeswitch.

**Comment by gafter** [comment]
@gulshan Can you please provide the proposed translation of this construct, as well as some realistic examples that illustrate the need this fulfils?

**Comment by alrz** [comment]
I think it's only possible to omit the type argument when it's known to the compiler.

**Comment by weitzhandler** [comment]
Named parameters or omission by index looks fine to me.
But I vote for the general idea of including non-inferred types only.
